File: 2024/11/11.py
# initial_state = [125, 17]

def parse(filename):
    with open(filename, 'r') as f:
        lines = f.readlines()
    stones = []
    for line in lines:
        _ = line.strip().split()
        for stone in _:
            stones.append(int(stone))
    return stones

def has_zero(stone):
    if stone == 0:
        return True
    return False

def is_even(stone):
    i = 0
    while(stone > 0):
        i += 1
        stone //= 10
    return i % 2 == 0, i

def partition(stone, size):
    div = 10 ** (size // 2)
    return stone // div, stone % div

def blink(number, initial_state):
    temp_state = []
    for i in range(number):
        for stone in initial_state:
            if has_zero(stone):
                temp_state.append(1)
            else:
                is_even_, length = is_even(stone)
                if is_even_:
                    left, right = partition(stone, length)
                    temp_state.append(left)
                    temp_state.append(right)
                else:
                    temp_state.append(stone * 2024)
        initial_state = temp_state
    return initial_state


if __name__ == '__main__':
    stones = parse("2024/11/11.txt")
    print(stones)
    print(blink(5, stones))

File: 2024/11/11.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"strconv"
	"strings"
)

 func parse(path string) []int64 {
	file, err := os.Open(path)
	if err != nil {
		log.Fatal("Error occured while loading the file", err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	var stones []int64
	for scanner.Scan() {
		text := scanner.Text()
		values := strings.Split(text, " ")
		for _, val := range values {
			i, err := strconv.ParseInt(val, 10, 64)
			if err != nil {
				log.Fatal("Error ocurred while transforming string -> int", err)
			}
			stones = append(stones, i)
		}
	}
	return stones
 }

 type Even struct {
	even bool
	size int
}

 func isEven(stone int64) Even {
	i := 0
	for stone > 0 {
		i += 1
		stone /= 10
	}
	return Even{i % 2 == 0, i}
 }

 type Partitions struct {
	left int64
	right int64
 }

 func partition(stone, size int64) Partitions {
	div := math.Pow(10, float64(size) / 2)
	left := stone / int64(div)
	right := stone % int64(div)
	return Partitions{left, right}
 }

 func blink(times int64, stones []int64) int64 {
	
 }

 func main() {
	stones := parse("/home/minindu/AoC/2024/11/11.txt")
	fmt.Println("Stones: ", stones)
 }